File: content/snippets/python/language_basics/recursion.py
#!/usr/bin/env python3

def factorial(n: int) -> int:
    if n <= 1:
        return 1
    return n * factorial(n - 1)


def sum_recursive(values: list[int], index: int = 0) -> int:
    if index >= len(values):
        return 0
    return values[index] + sum_recursive(values, index + 1)


def main() -> None:
    values = [1, 2, 3, 4]
    print("factorial(5):", factorial(5))
    print("sum_recursive(values):", sum_recursive(values))


if __name__ == "__main__":
    main()

File: content/snippets/typescript/language_basics/recursion.ts
#!/usr/bin/env ts-node

import { setImmediate } from "node:timers";

function factorial(n: number): number {
  if (n <= 1) {
    return 1;
  }
  return n * factorial(n - 1);
}

function sumRecursive(values: number[], index = 0): number {
  if (index >= values.length) {
    return 0;
  }
  return values[index] + sumRecursive(values, index + 1);
}

function main(): void {
  const values = [1, 2, 3, 4];
  console.log("factorial(5):", factorial(5));
  console.log("sumRecursive(values):", sumRecursive(values));
}

main();
setImmediate(() => {});
